File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(list1 *ListNode, list2 *ListNode) *ListNode {
    if list1 == nil {
        return list2
    }
    if list2 == nil {
        return list1
    }
    
    var dummy = &ListNode{}
    var curr = dummy
    
    for list1 != nil && list2 != nil {
        if list1.Val < list2.Val {
            curr.Next = list1
            list1 = list1.Next
        } else {
            curr.Next = list2
            list2 = list2.Next
        }
        curr = curr.Next
    }
    
    if list1 != nil {
        curr.Next = list1
    }
    if list2 != nil {
        curr.Next = list2
    }
    
    return dummy.Next
}

File: 0021-merge-two-sorted-lists/0021-merge-two-sorted-lists.rs
impl Solution {
    pub fn merge_two_lists(mut list1: Option<Box<ListNode>>, mut list2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
        let mut dummy = Box::new(ListNode::new(0));
        let mut tail = &mut dummy;

        while list1.is_some() && list2.is_some() {
            if list1.as_ref().unwrap().val < list2.as_ref().unwrap().val {
                let next = list1.as_mut().unwrap().next.take();
                tail.next = list1.take();
                list1 = next;
            } else {
                let next = list2.as_mut().unwrap().next.take();
                tail.next = list2.take();
                list2 = next;
            }
            tail = tail.next.as_mut().unwrap();
        }

        if list1.is_some() {
            tail.next = list1;
        } else {
            tail.next = list2;
        }

        dummy.next
    }
}
